File: Intermediate/python/StringEvolution.py
import random
import string
import time

# --- Configuration ---
TARGET = "BackupCookie686"
POPULATION_SIZE = 100
MUTATION_RATE = 0.01  # 1% chance per character

# --- Helper Functions ---
def get_random_char():
    """Returns a random printable character."""
    pool = string.ascii_letters + string.digits + " !.,;:%@#$^&*()"
    return random.choice(pool)

def create_random_string(length):
    """Generates a random string of a specific length."""
    return "".join(get_random_char() for _ in range(length))

def calculate_fitness(individual, target):
    """Counts how many characters match the target exactly."""
    return sum(1 for c, t in zip(individual, target) if c == t)

def mutate(individual):
    """Randomly alters characters based on mutation rate."""
    genes = list(individual)
    for i in range(len(genes)):
        if random.random() < MUTATION_RATE:
            genes[i] = get_random_char()
    return "".join(genes)

# --- Main Evolution Loop ---
def run_evolution():
    # 1. Initialize random population
    population = [create_random_string(len(TARGET)) for _ in range(POPULATION_SIZE)]
    generation = 0
    
    while True:
        generation += 1
        
        # 2. Evaluate fitness for everyone
        scored_population = [(calculate_fitness(ind, TARGET), ind) for ind in population]
        
        # 3. Sort by highest fitness first
        scored_population.sort(key=lambda x: x[0], reverse=True)
        best_fitness, best_individual = scored_population[0]
        
        # Print progress
        print(f"Gen {generation:03d} | Best: '{best_individual}' | Fitness: {best_fitness}/{len(TARGET)}")
        
        # 4. Check for success condition
        if best_individual == TARGET:
            print(f"\nSuccess! Target reached in {generation} generations.")
            break
            
        # 5. Breed the next generation using the best individual (Elitist cloning)
        next_generation = [best_individual] # Keep the absolute best
        
        while len(next_generation) < POPULATION_SIZE:
            # Create variations of the best individual
            offspring = mutate(best_individual)
            next_generation.append(offspring)
            
        population = next_generation
        
def speedtest():
    timings = []
    for i in range(100):
        start = time.time()*1000
        run_evolution()
        end = time.time()*1000
        timings.append(int(end-start))
        
    timings.sort()
    
    print(timings[0])
    print(timings[1])

if __name__ == "__main__":
    # start = time.time()*1000
    run_evolution()
    # end = time.time()*1000
    
    # print (start)
    # print (end)
    
    # print(int(end-start))
    
    # speedtest()

File: Intermediate/Advance C++/StringEvolution.cpp
#include <iostream>
#include <array>
#include <string>
#include <random>
#include <time.h>
#include <tuple>
using namespace std;

static string TARGET = "BackupCookie686";
constexpr int POPULATION_SIZE = 100;
constexpr double MUTATION_RATE = 0.01;

typedef array<string, POPULATION_SIZE> populationArray;
typedef array<tuple<int, string>, POPULATION_SIZE> score_population_Array;

char get_random_char()
{
    string pool = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 !.,;:%@#$^&*()";
    return pool[rand() % (int)pool.size()];
}

string create_random_string(int length)
{
    string ret;
    for (int i = 0; i < length; i++)
        ret += get_random_char();
    return ret;
}

void populate(populationArray& population)
{
    for (int i = 0; i < POPULATION_SIZE; i++)
    {
        population[i] = create_random_string(TARGET.length());
    }
}

int calculate_fitness(const string& ind, const string& target)
{
    return 1;
}

void calculate_total_score(score_population_Array& scorearr, const populationArray& poparr)
{
    for (int i = 0; i < POPULATION_SIZE; i++)
        scorearr[i] = make_tuple(calculate_fitness(poparr[i], TARGET), poparr[i]);
}

void displayPopulation(const populationArray& pop)
{
    for (int i = 0; i< POPULATION_SIZE; i++)
        cout << pop[i] << "\n";
}

void displayScorePopulation(score_population_Array arr)
{
    // for (auto& [score, ind] : arr)
    // {
    //     cout << score << " " << ind << "\n";
    // }
    // for (auto e : arr)
    // {
    //     cout << e. << " " << ind << "\n";
    // }
}

void run_evolution()
{
    populationArray population;
    populate(population);
    int generation = 0;

    while (true)
    {
        generation += 1;
        // pair (score, ind)
        score_population_Array scored_population;
        calculate_total_score(scored_population, population);
        displayScorePopulation(scored_population);
        break;
    }
    
    // displayPopulation(population);
}


int main()
{
    srand(time(0));
    // cout << create_random_string(12);
    // for (int i = 0; i < 10000; i++)
    //     cout << get_random_char() << "\n";
    run_evolution();
}